sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,465 @@
+ * To configure the JDBC source, you have to provide a {@link DataSource}. The username and
+ * password to connect to the database are optionals. The following example illustrates how to
+ * configure a JDBC source:
+ * </p>
+ * <pre>
+ *   {@code
+ *
+ * pipeline.apply(JdbcIO.read()
+ *   .withDataSource(myDataSource)
+ *   .withRowMapper(new JdbcIO.RowMapper<MyElement>() {
+ *     public MyElement mapRow(ResultSet resultSet) {
+ *       // use the resultSet to build the element of the PCollection
+ *       // for instance:
+ *       // return resultSet.getString(2);
+ *     }
+ *   })
+ *
+ *   }
+ * </pre>
+ * <p>
+ *   You can find an full {@link RowMapper} example in {@code JdbcIOTest}.
+ * </p>
+ * <h3>Writing to JDBC datasource</h3>
+ * <p>
+ * JDBC sink supports writing records into a database. It expects a
+ * {@code PCollection<T>}, converts the {@code T} elements as SQL statement
+ * and setParameters into the database. T is the type expected by the
+ * provided {@link PreparedStatementSetter}.
+ * </p>
+ * <p>
+ * Like the source, to configure JDBC sink, you have to provide a datasource. For instance:
+ * </p>
+ * <pre>
+ *   {@code
+ *
+ * pipeline
+ *   .apply(...)
+ *   .apply(JdbcIO.write()
+ *      .withDataSource(myDataSource)
+ *      .withQuery(query)
+ *      .withPreparedStatementSetter(new JdbcIO.PreparedStatementSetter<MyElement>() {
+ *        public void setParameters(MyElement element, PreparedStatement statement) {
+ *          // use the PCollection element to set parameters of the SQL statement used to insert
+ *          // in the database
+ *          // for instance: statement.setString(0, element.toString());
+ *        }
+ *      })
+ *
+ *   }
+ * </pre>
+ * <p>
+ *   You can find a full {@link PreparedStatementSetter} in {@code JdbcIOTest}.
+ * </p>
+ */
+public class JdbcIO {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(JdbcIO.class);
+
+  /**
+   * Read data from a JDBC datasource.
+   *
+   * @return a {@link Read} {@link PTransform}.
+   */
+  public static Read<?> read() {
+    return new Read(new Read.JdbcOptions(null, null, null, null), null);
+  }
+
+  /**
+   * Write data to a JDBC datasource.
+   *
+   * @return a {@link Write} {@link PTransform}.
+   */
+  public static Write<?> write() {
+    return new Write(new Write.JdbcWriter(null, null, null, null,
+        null, 1024L));
+  }
+
+  private JdbcIO() {
+  }
+
+  /**
+   * An interface used by the JdbcIO Read for mapping rows of a ResultSet on a per-row basis.
+   * Implementations of this interface perform the actual work of mapping each row to a result
+   * object used in the {@link PCollection}.
+   */
+  public interface RowMapper<T> extends Serializable {
+      T mapRow(ResultSet resultSet);
+  }
+
+  /**
+   * A {@link PTransform} to read data from a JDBC datasource.
+   */
+  public static class Read<T> extends PTransform<PBegin, PCollection<T>> {
+
+    public Read<T> withDataSource(DataSource dataSource) {
```

Comment:
Just noticed that DataSource does not extend Serializable. You might need to replace it with a custom interface for creating connections inside JdbcIO. You could call it ConnectionFactory or something.